File: Practice/atcoder/ABC/032/src/c.py
# いろいろ試したんだけどコツは
# - 尺取法
# - 積は計算するのではなく結果をもっておいて更新していく
# - なんかr==lでr+=1とseki=1する <- 意味がわからん

n, k = map(int, input().split())
arr = [int(input()) for _ in range(n)]
if 0 in arr:
    print(n)
    exit()

length = 0
seki = 1
r = 0
for i in range(n):
    # l = arr[i]
    # r = max(i, r)
    # print("l: {} r: {}".format(i, r))
    # while r < n-1 and seki*arr[r] <= k:
    while r < n and seki*arr[r] <= k:
        # print("seki: {} arr[{}]: ".format(seki, arr[r]))
        seki *= arr[r]
        # print("now seki {}".format(seki))
        r += 1

    length = max(length, r - i)
    #  seki //= l
    seki /= arr[i]
    # seki = max(1, seki)  # 場当たり的す
    # ぎる
    if r == i:  # ????????????????????????
        r += 1
        seki = 1
    # print("{}".format(arr[i:r - i]))
    # print("length: {}".format(length))

print(length)

File: Practice/atcoder/ABC/032/src/_c.php
完全に意味不明なんだけどなぜかWAなる
<?php
  // 尺取典型
  $nk = array_map("intVal", explode(" ", trim(fgets(STDIN))));
  $n = $nk[0];
  $k = $nk[1];
  $arr = [];
  for($i=0; $i<$n; $i++){
    $t = intVal(trim(fgets(STDIN)));
    if($t == 0){
      echo $n . PHP_EOL;
      exit();
    }
    $arr[] = $t;
  }
  $l = 0;
  $r = 0;
  $ans = 0;
  $tmp = 1;
  while($r < $n-1){
    if($tmp * $arr[$r+1] <= $k){
      $tmp *= $arr[$r+1];
      $r += 1;
      // ansは長さなので
      $ans = max($ans, $r - $l);
    }elseif($r == $l){
      $r += 1;
      $l += 1;
    }else{
      $tmp /= $arr[$l];
      $l += 1;
    }
  }
  echo $ans . PHP_EOL;
?>